### DISCUSSION TITLE
gfx1100 (RX 7900 XTX): rocBLAS f16 GEMM intermittently returns whole NaN output columns. Reproducer inside; looking for other cards to compare

### DISCUSSION THREAD

**Discussion by roverdrubber** [ROCm/TheRock#8355] [Q&A]
Hi TheRock team and users. I'm trying to find out whether a silent-corruption problem is a fault in my particular RX 7900 XTX or something in the stack, and I'd appreciate your guidance, plus data from other gfx1100 cards. I'm not assuming a ROCm bug: other cards passing would point at my card.

### System

- Gigabyte RX 7900 XTX 24 GB, VBIOS `113-EXT780190-001` (OC position) and `113-EXT80189-001` (Silent position), PCI `1002:744c`, subsystem `1458:240e`
- Ryzen 7 5700X, 96 GB DDR4, Ubuntu 26.04, kernel `7.0.0-31-generic`, in-tree amdgpu
- TheRock ROCm 7.14.1 tarball: package `7.14.1rc0`, TheRock commit `f51dc6c91e0d`, CI run `32200708233`; rocm-libraries `cd9574023093`, rocm-systems `ca887ee80abf`, llvm-project `f30ae3e6b680` (also tested with ROCm 7.2.4 and 10.0)

### Reproducer

Source, build line, a 10-second self-test and what to report: [https://gist.github.com/roverdrubber/7cd347a9340abdad2f707f2252aaaee0](https://gist.github.com/roverdrubber/7cd347a9340abdad2f707f2252aaaee0)

```bash
export ROCM_PATH=/path/to/your/rocm      # required: compiler and rocBLAS from the same installation
"$ROCM_PATH/bin/hipcc" -O2 --offload-arch=gfx1100 gemmcheck3.hip -o gemmcheck3 \
    -L"$ROCM_PATH/lib" -lrocblas -Wl,-rpath,"$ROCM_PATH/lib"
GEMMCHECK_INJECT=7 ./gemmcheck3 10 5      # self-test: must report the planted 64x16 fault, exit 1
GEMMCHECK_STOP=1 ./gemmcheck3 600 5       # the real run: approximately 10 minutes, checked at ~5 s reporting checkpoints
```

It runs `rocblas_gemm_ex` with `rocblas_atomics_not_allowed`: five GEMMs in LLM layer shapes (fp16 in / fp32 out: 17408x512x5120 and 5120x512x17408; int8 in / int32 out: the same plus 12288x512x5120), random inputs, back to back in one stream. Each output is compared bit for bit on the GPU against a reference computed at startup, after a determinism check passes. On a mismatch, the compare kernel records the launch, index, value read, reference, and both re-read through an atomic (served by L2).

### What it catches on my card (stock clocks, ~2460 MHz, 300 W, hotspot 76 to 81 °C)

- Every event captured in full detail so far (three, all on 7.14.1) is one or more **complete output columns** (all 17,408 rows) of the fp16 17408x512x5120 GEMM becoming **nonfinite**. The printed samples were NaN with one payload per event (`0xFFFFE000`, the f32 of f16 `0xFFFF`; `0xFFE00000`, the f32 of f16 `0xFF00`); uniform payloads across a whole column weren't verified.
- The atomic re-read returns the same wrong values, which strongly supports corrupted GEMM output rather than a transient checker-cache misread. The column pattern fits a bad read of the shared B column, but that's a hypothesis.
- Bursty: two early 20-minute runs each had three reporting intervals with new mismatches; some 20 to 30 minute runs were clean. The latest was detected 397 s after a cold start.
- The detailed captures are from 7.14.1; 7.2.4 also produced mismatch counts; one 20-minute run on 10.0 was clean.
- After about 2.5 hours of sustained load the errors escalated, also in the int8 GEMMs, and flashing artifacts appeared on the desktop, clearing a minute or two after the load stopped.
- No amdgpu fault or reset lines during these GEMM runs. Separately, once during llama.cpp inference, the card logged address-zero page faults followed by MES failures and dropped off the PCIe bus, which resembles ROCm#6529 (a different workload; no shared root cause established).

rocprofv3 on 7.14.1 shows one dispatch per GEMM call. The fp16 GEMMs use a generic fallback kernel (`ISA000`, `GSU1`, no split-K; the name indicates wave64, `WS64`); the int8 ones use WMMA kernels (`MI16x16x16x1`, `ISA1100`, wave32, `WS32`).

<details>
<summary>Full kernel names (rocprofv3 kernel trace, ROCm 7.14.1)</summary>

```
fp16 (both shapes):
Cijk_Alik_Bljk_HSS_BH_MT64x32x8_SE_1LDSB0_AMAS2_BL1_BS1_EPS0_GLVWA2_GLVWB1_GRVW2_GSU1_GSUASB_ISA000_IU1_K1_KLS_LBSPPA0_LBSPPB0_LPA0_LPB0_LRVW2_MIAV0_MMFSC_NLCA1_NLCB1_PGR0_PLR1_SIA1_SS0_SU32_SUM0_SUS256_SVW4_TT4_2_TLDS0_UMLDSA0_UMLDSB0_USFGROn1_VAW2_VSn1_VW2_VWB2_WSGRA0_WSGRB0_WS64_WG16_16_1_WGM8

int8:
Cijk_Alik_Bljk_I8II_BH_MT128x128x32_MI16x16x16x1_SN_1LDSB0_AMAS0_BL1_BS1_EPS1_GLVWA8_GLVWB8_GRVW8_GSU1_GSUASB_ISA1100_IU1_K1_KLA_LBSPPA128_LBSPPB128_LPA16_LPB16_LRVW16_MIAV1_MMFGLC_NLCA1_NLCB1_PGR1_PLR1_SIA3_SS1_SU0_SUM0_SUS0_SVW1_TT4_64_TLDS1_UMLDSA1_UMLDSB1_USFGROn1_VAW1_VSn1_VW1_VWB4_WSGRA1_WSGRB1_WS32_WG32_4_1_WGM8
Cijk_Alik_Bljk_I8II_BH_MT128x128x32_MI16x16x16x1_SN_1LDSB0_AMAS3_BL1_BS1_EPS1_GLVWA16_GLVWB16_GRVW16_GSU1_GSUASB_ISA1100_IU1_K1_KLA_LBSPPA128_LBSPPB128_LPA16_LPB16_LRVW16_MIAV1_MMFGLC_NLCA1_NLCB1_PGR1_PLR1_SIA3_SS1_SU0_SUM0_SUS0_SVW2_TT4_64_TLDS1_UMLDSA1_UMLDSB1_USFGROn1_VAW1_VSn1_VW2_VWB4_WSGRA1_WSGRB1_WS32_WG32_4_1_WGM4
```

</details>

<details>
<summary>How I got here, and what I've already varied</summary>

I noticed it because llama.cpp on this card degenerates into `////` / NaN logits on long prompts (roughly 25 to 60 percent of fresh server processes at 17.5K tokens), with the first observed divergence at a quantized matmul or the normalization after it, NaN payloads from the same family, and the weights in VRAM byte-identical to the file.

In llama.cpp trials (not all of these were repeated with the standalone GEMM test), none of the following eliminated the corruption: ROCm 7.1, 7.2.4, 7.14.1 and 10.0; AMD's DKMS driver 7.1.3 and the in-tree driver; kernels 6.17 and 7.0; four GC firmware builds; a 24.04 userspace; Windows (HIP SDK); several llama.cpp builds including AMD's prebuilt b8407; two PSUs; a repaste; both BIOS switch positions. Diagnostic passes: memtest86+ on system RAM. Capping the core clock and `HIP_LAUNCH_BLOCKING=1` reduce it in llama.cpp but don't fix it. OCCT VRAM and 3D (error detection on, 30 to 40 minutes each) and memtest_vulkan pass on this card. The previous card in the same PC, an RX 6900 XT, never had this.

</details>

### Questions

1. **Is wrong output from this HSS fallback kernel (or GEMM on gfx1100 generally) a known issue?** Does AMD's CI run long, bit-exact GEMM repetition on gfx1100 that would catch this?
2. **How can I pin a different HSS solution** (a solution index, or an environment variable) to see whether the fault follows this kernel?
3. **Would results from the ROCm Validation Suite** (for example `gst` or the memory tests) be more useful to you than my reproducer? If so, which modules and settings?
4. **Is anything between 7.14.1 and 10.0 known to affect gfx1100 GEMM correctness?** One clean 20-minute run on 10.0 isn't evidence of a fix. If you point me to a build worth trying, I can run bounded, supervised tests on it that stop at the first mismatch or artifact.
5. **If you have a gfx1100 card,** would you run the 10-minute check and post your ROCm version, kernel, clock/voltage settings, elapsed time, launch count and the last lines of the output? It's a heavy load: please run it supervised and stop at any artifacts.

Clean runs from other cards would strengthen the suspicion that it's my particular card; similar failures elsewhere would point to a shared issue, though not prove which. Thank you for any help.

**Comment by roverdrubber** [comment]
**Update after two more days of testing. Filed as a rocBLAS issue with the details: [ROCm/rocm-libraries#12344*](https://github.com/ROCm/rocm-libraries/issues/12344)*

Two corrections to my post above. The clean 20-minute run on ROCm 10.0 did not hold up: 10.0 faulted on a later run (first mismatch at 188 s). And I described the escalation after 2.5 hours of load as a progression; I no longer think the data supports that, see the last point.

What is new:

- **Another owner reports the same thing on a different RX 7900 XTX** (their output is in the comments of [this r/ROCm thread](https://www.reddit.com/r/ROCm/comments/1wjxvuc/gfx1100_rx_7900_xtx_rocblas_f16_gemm/); Debian 13, kernel 6.12.107, ROCm 7.14). Two runs of the unmodified reproducer, two mismatches, both in the fp16 17408x512x5120 case, both one output row wrong across columns 0-255. That is their report, not something I verified; my card produces the same shape.
- **On my card the reproducer has now faulted with four ROCm userspaces:** 6.4.4 (container, on the same host 7.0 kernel driver), 7.2.4, 7.14.1 and 10.0. So it is not confined to 7.x userspace.
- **Controls:** four HIP kernels of mine that call no library (FMA chains, WMMA instructions issued directly, two tiled matmuls) each ran 10 minutes without a mismatch under the same checker, each bracketed by rocBLAS runs that mismatched within 6 to 152 s. They differ from the rocBLAS kernels in instruction mix, memory access and throughput, so this does not isolate the cause.
- **Geometry on my card** (25 fp16 and 14 int8 events captured): every fp16 single-row event, 10 of 10, covers exactly columns 0-255, also with n raised to 1024; fp16 whole-column events have all been below column 256; int8 events are not confined.
- `HIP_LAUNCH_BLOCKING=1` did not eliminate it (4 events in 165,120 calls with it, 3 in 167,040 without).
- **It is bursty and I found no reliable predictor.** Same binary and similar clock, power and temperature: first mismatch anywhere from 6 s to 397 s, other 10 to 30 minute runs clean. One boot on 2026-09-20: clean 600 s, mismatch at 12 s, clean 600 s.

I still cannot tell a software defect from a hardware margin. Questions 2 and 4 in my original post are superseded by the issue.

**Comment by roverdrubber** [comment]
**Resolved, and it was my hardware. Withdrawing the suggestion that this is a ROCm or rocBLAS defect.**

I swapped the RX 7900 XTX for an RX 6900 XT and reran everything on the same board, PSU, slot, kernel,
ROCm 7.14.1 and the same binary, built fat for `gfx1030;gfx1100` so both cards ran the same executable:

| | RX 7900 XTX | RX 6900 XT |
|---|---|---|
| `gemmcheck3` reproducer, 600 s | first mismatch at 6-69 s | `RESULT CLEAN: 87360 GEMM launches, 0 mismatched words` |
| llama.cpp fresh-server starts | 32 of 104 corrupted (~31%) | 0 of 34 |

Four other owners have since run the same reproducer on five other RX 7900 XTX / W7900 cards and all
came back clean, including one that is the sibling of a faulting card in the *same machine* -- which no
library bug can distinguish -- and one on Windows. Turning XMP and PBO off changed nothing on mine, so
the board, CPU, four-DIMM population, PSU, slot and kernel are all cleared too.

Two corrections to my earlier posts while I am here: the clean 20-minute run I reported on ROCm 10.0 did
not hold up (10.0 faulted at 188 s on a later run), and I described an escalation after 2.5 hours of
load as a progression, which the data does not support -- the fault is bursty, with the first mismatch
falling anywhere from 6 s to 397 s and other 10 to 30 minute runs clean.

Sorry for the noise. The one thing I would still flag, and it is not a TheRock issue: a marginal card
returns silently wrong GEMM results with no error anywhere -- no HSA exception, no `dmesg` line, nothing
-- and `rocm-smi` and ROCmValidationSuite both pass it. That is what made this take two weeks.
